openpgp/packet/aead_encrypted.go

```diff
@@ -0,0 +1,408 @@
+		return nil, errors.AEADError("Not enough data to decrypt")
+	}
+	return &streamReader{
+		worker: worker{
+			aead:         aead,
+			config:       config,
+			initialNonce: ae.initialNonce,
+			header:       ae.prefix,
+			index:        make([]byte, 8),
+		},
+		reader: ae.Contents,
+		carry:  carry}, nil
+}
+
+// Read decrypts bytes and reads them into dst. It decrypts when necessary and
+// caches extra decrypted bytes. It returns the number of bytes copied into dst
+// and an error.
+func (ar *streamReader) Read(dst []byte) (n int, err error) {
+	if len(dst) == 0 {
+		return 0, errors.AEADError("argument of Read must have positive length")
+	}
+	chunkLen := int(ar.config.ChunkSize())
+	tagLen := ar.config.TagLength()
+	if len(dst) <= len(ar.cache) {
+		n = copy(dst, ar.cache[:len(dst)])
+		ar.cache = ar.cache[n:]
+		return
+	}
+	// Retrieve cached plaintext bytes from previous calls
+	decrypted := ar.cache
+
+	// Decrypt the necessary amount of chunks
+	for i := 0; i <= (len(dst)-len(ar.cache))/chunkLen; i++ {
+		cipherChunk := make([]byte, chunkLen+tagLen)
+		readBytes, err := io.ReadFull(ar.reader, cipherChunk)
+		if err != nil && err != io.EOF && err != io.ErrUnexpectedEOF {
+			return 0, err
+		}
+		err = err
+		if readBytes == 0 {
+			break
+		}
+		cipherChunk = cipherChunk[:readBytes]
+		plainChunk, errChunk := ar.processChunk(cipherChunk)
+		if errChunk != nil {
+			return 0, errChunk
+		}
+		decrypted = append(decrypted, plainChunk...)
+		if err == io.EOF || err == io.ErrUnexpectedEOF {
+			break
+		}
+	}
+	// Append necessary bytes, and cache the rest
+	if len(dst) < len(decrypted) {
+		n = copy(dst, decrypted[:len(dst)])
+		ar.cache = decrypted[len(dst):]
+	} else {
+		n = copy(dst, decrypted)
+		ar.cache = nil
+	}
+	return
+
+}
+
+// Close wipes the worker, along with the reader, cached, and carried bytes.
+func (ar *streamReader) Close() (err error) {
+	ar.worker = worker{}
+	ar.carry = nil
+	return nil
+}
+
+// GetStreamWriter initializes the worker and returns a writer. This writer
+// encrypts and writes bytes (see streamWriter.Write()).
+func GetStreamWriter(w io.Writer, key []byte, config *AEADConfig) (io.WriteCloser, error) {
+	writeCloser := noOpCloser{w}
+	writer, err := serializeStreamHeader(writeCloser, packetTypeAEADEncrypted)
+	if err != nil {
+		return nil, err
+	}
+	// Set packet tag
+	buf := bytes.NewBuffer(nil)
+	if err := serializeType(buf, packetTypeAEADEncrypted); err != nil {
+		return nil, err
+	}
+	// Data for en/decryption: tag, version, cipher, aead mode, chunk size
+	header := []byte{
+		config.Version(),
+		byte(config.Cipher()),
+		byte(config.Mode()),
+		config.ChunkSizeByte()}
+	n, err := writer.Write(header)
+	if err != nil || n < 4 {
+		return nil, errors.AEADError("could not write AEAD headers")
+	}
+	// Sample nonce
+	nonce := make([]byte, config.NonceLength())
+	n, err = rand.Read(nonce)
+	if n < config.NonceLength() || err != nil {
+		panic("Could not sample random nonce")
+	}
```

Comment:
If you switch to `ReadFull`, the first half of this `if` condition shouldn't be necessary.